kdc/kerberos5.c

```diff
@@ -2432,12 +2432,29 @@ _kdc_tkt_add_if_relevant_ad(krb5_context context,
 }
 
 krb5_boolean
-_kdc_is_anon_request(const KDC_REQ_BODY *b)
+_kdc_is_anon_request(KDC_REQ_BODY *b)
 {
-	/* some versions of heimdal use bit 14 instead of 16 for
-	   request_anonymous, as indicated in the anonymous draft prior to
-	   version 11. Bit 14 is assigned to S4U2Proxy, but all S4U2Proxy
-	   requests will have a second ticket; don't consider those anonymous */
-	return (b->kdc_options.request_anonymous ||
-		(b->kdc_options.constrained_delegation && !b->additional_tickets));
+    if (b->kdc_options.constrained_delegation) {
+	if (!b->additional_tickets) {
+	    /*
+	     * some versions of heimdal use bit 14 instead of 16 for
+	     * request_anonymous, as indicated in the anonymous draft prior
+	     * to version 11. Bit 14 is assigned to S4U2Proxy, but all
+	     * S4U2Proxy requests will have a second ticket.  Use the missing
+	     * second ticket as an indicator that this is a pre-draft 11
+	     * client. */
+	    b->kdc_options.constrained_delegation = 0;
+	    b->kdc_options.request_anonymous = 1;
+	}
+	else if (b->kdc_options.request_anonymous) {
+	    /*
+	     * Heimdal versions between 1.0 and 7.6 set the anonymous_request
+	     * bit (16) in addition to constrained_delegation (14) when
+	     * requesting S4U2Proxy.  The Windows KDC ignores the
+	     * anonymous_request bit when satisfying a S4U2Proxy request.
+	     */
+	    b->kdc_options.request_anonymous = 0;
```

Comment:
So here we think it is a simple S4U2Proxy request wrongly flagged anonymous, are we sure _kdc_is_anon_request() is called from TGS? some of the calls seem to be from AS request (also, does it make sense to check if the tgt is anon?).
I think I would prefer to see this fixup in more specific location in the flow.

Reply:
This change only impacts AS requests and not TGS request.   

kdc/krb5tgs.c check_tgs_flags() might be a good location to perform these checks.  Its called at the start of tgs_make_reply().

Reply:
I will look at this again on Monday

Reply:
Is there an approach that doesn't involve rewriting the body flags? This feels wrong to me.

Reply:
BTW I committed the cname_in_addl_tkt change to master.

Reply:
Thanks.